services/src/layers/postgres_layer_db.rs

```diff
@@ -617,6 +619,216 @@ where
+                JOIN (SELECT DISTINCT layer FROM collection_layers JOIN parents ON (collection = id)) cl ON (id = cl.layer)
+            WHERE name ILIKE $4
+        )
+        ORDER BY is_layer ASC, name ASC
+        LIMIT $2 
+        OFFSET $3;           
+        ",
+            )
+            .await?;
+
+        let rows = conn
+            .query(
+                &stmt,
+                &[
+                    &collection,
+                    &i64::from(search.limit),
+                    &i64::from(search.offset),
+                    &pattern,
+                ],
+            )
+            .await?;
+
+        let items = rows
+            .into_iter()
+            .map(|row| {
+                let is_layer: bool = row.get(4);
+
+                if is_layer {
+                    Ok(CollectionItem::Layer(LayerListing {
+                        id: ProviderLayerId {
+                            provider_id: INTERNAL_PROVIDER_ID,
+                            layer_id: LayerId(row.get(0)),
+                        },
+                        name: row.get(1),
+                        description: row.get(2),
+                        properties: row.get(3),
+                    }))
+                } else {
+                    Ok(CollectionItem::Collection(LayerCollectionListing {
+                        id: ProviderLayerCollectionId {
+                            provider_id: INTERNAL_PROVIDER_ID,
+                            collection_id: LayerCollectionId(row.get(0)),
+                        },
+                        name: row.get(1),
+                        description: row.get(2),
+                        properties: row.get(3),
+                    }))
+                }
+            })
+            .collect::<Result<Vec<CollectionItem>>>()?;
+
+        Ok(LayerCollection {
+            id: ProviderLayerCollectionId {
+                provider_id: INTERNAL_PROVIDER_ID,
+                collection_id: collection_id.clone(),
+            },
+            name,
+            description,
+            items,
+            entry_label: None,
+            properties,
+        })
+    }
+
+    #[allow(clippy::too_many_lines)]
+    async fn autocomplete_search(
+        &self,
+        collection_id: &LayerCollectionId,
+        search: SearchParameters,
+    ) -> Result<Vec<String>> {
+        let collection = Uuid::from_str(&collection_id.0).map_err(|_| {
+            crate::error::Error::IdStringMustBeUuid {
+                found: collection_id.0.clone(),
+            }
+        })?;
+
+        let conn = self.conn_pool.get().await?;
+
+        let pattern = match search.search_type {
+            SearchType::Fulltext => {
+                format!("%{}%", search.search_string)
+            }
+            SearchType::Prefix => {
+                format!("{}%", search.search_string)
+            }
+        };
+
+        let stmt = conn
+            .prepare(
+                "
+        WITH RECURSIVE parents AS (
```

Comment:
re-use the query?

Reply:
https://github.com/geo-engine/geoengine/pull/908/commits/5b36378fe3c8ee1e601a86007dc0ea58fb56df75